File: java8/src/main/java/folio/codinginterview/application/usecase/portfolio/UpdatePortfolioUsecase.java
package folio.codinginterview.application.usecase.portfolio;

import folio.codinginterview.application.repository.PortfolioRepository;
import folio.codinginterview.domain.StockSymbol;
import folio.codinginterview.domain.Portfolio;
import folio.codinginterview.domain.PortfolioItem;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public final class UpdatePortfolioUsecase {
    public static final class ItemInput {
        private final StockSymbol symbol;
        private final BigDecimal rate;
        public ItemInput(StockSymbol symbol, BigDecimal rate) {
            this.symbol = symbol;
            this.rate = rate;
        }
        public StockSymbol symbol() { return symbol; }
        public BigDecimal rate() { return rate; }
    }

    public static final class Input {
        private final List<ItemInput> items;
        public Input(List<ItemInput> items) {
            this.items = Collections.unmodifiableList(new ArrayList<>(items));
        }
        public List<ItemInput> items() { return items; }
    }

    public static class Exception extends RuntimeException {
        protected Exception(String msg) { super(msg); }
    }

    public static final class InvalidPortfolio extends Exception {
        public InvalidPortfolio(String reason) { super(reason); }
    }

    private final PortfolioRepository portfolioRepository;

    public UpdatePortfolioUsecase(PortfolioRepository portfolioRepository) {
        this.portfolioRepository = portfolioRepository;
    }

    public CompletableFuture<Void> run(Input input) {
        Portfolio portfolio;
        try {
            List<PortfolioItem> items = new ArrayList<>();
            for (ItemInput i : input.items()) {
                items.add(new PortfolioItem(i.symbol(), i.rate()));
            }
            portfolio = new Portfolio(items);
        } catch (RuntimeException e) {
            CompletableFuture<Void> failed = new CompletableFuture<>();
            failed.completeExceptionally(new InvalidPortfolio(e.getMessage()));
            return failed;
        }
        return portfolioRepository.update(portfolio);
    }
}

File: php/src/CodingInterview/Application/Usecase/Portfolio/UpdatePortfolioUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Portfolio;

use Folio\CodingInterview\Application\Repository\PortfolioRepository;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\StockSymbol;
use Folio\CodingInterview\Domain\Portfolio;
use Folio\CodingInterview\Domain\PortfolioItem;

final class UpdatePortfolioItemInput
{
    public function __construct(
        public readonly StockSymbol $symbol,
        public readonly BigDecimal $rate,
    ) {}
}

final class UpdatePortfolioUsecaseInput
{
    /** @param UpdatePortfolioItemInput[] $items */
    public function __construct(public readonly array $items) {}
}

final class UpdatePortfolioInvalidPortfolioException extends \RuntimeException
{
}

final class UpdatePortfolioUsecase
{
    public function __construct(private readonly PortfolioRepository $portfolioRepository) {}

    public function run(UpdatePortfolioUsecaseInput $input): void
    {
        try {
            $items = [];
            foreach ($input->items as $i) {
                $items[] = new PortfolioItem($i->symbol, $i->rate);
            }
            $portfolio = new Portfolio($items);
        } catch (\Throwable $e) {
            throw new UpdatePortfolioInvalidPortfolioException($e->getMessage(), 0, $e);
        }
        $this->portfolioRepository->update($portfolio);
    }
}
